leases/tasks/send_invoices.py

```diff
@@ -0,0 +1,231 @@
+      (2) during or after season: leases from this year
+    """
+
+    current_date = today().date()
+    # If today is before the season starts but during the same year (1)
+    if current_date < season_start and current_date.year == season_start.year:
+        lease_year = current_date.year - 1
+    else:  # (2)
+        lease_year = current_date.year
+
+    # Filter leases from the upcoming season
+    future_leases = BerthLease.objects.filter(
+        start_date__year__gt=lease_year,
+        berth=OuterRef("berth"),
+        customer=OuterRef("customer"),
+    )
+
+    # Exclude leases that have already been assigned to the same customer and berth on the future
+    leases: QuerySet = BerthLease.objects.exclude(
+        Exists(future_leases.values("pk"))
+    ).filter(
+        # Only allow leases that are auto-renewing and have been paid
+        renew_automatically=True,
+        status=LeaseStatus.PAID,
+        start_date__year=lease_year,
+        end_date__year=lease_year,
+    )
+
+    return leases
+
+
+def create_new_lease(
+    lease: Union[BerthLease, WinterStorageLease], start_date: date, end_date: date
+) -> Union[BerthLease, WinterStorageLease]:
+    """Create a new lease instance"""
+
+    # Blank the PK to signal that a new instance has to be created
+    lease.pk = None
+
+    # Manually need to set creating because of the checks performed on clean
+    lease._state.adding = True
+
+    # Update the dates for the lease for the next season
+    lease.start_date = start_date
+    lease.end_date = end_date
+    lease.save()
+
+    return lease
+
+
+def send_berth_invoices(request: HttpRequest, due_date: date = None) -> Dict[str, List]:
+    return send_invoices(
+        request,
+        BERTH_LEASES,
+        season_start=calculate_season_start_date(),
+        season_end=calculate_season_end_date(),
+        due_date=due_date,
+    )
+
+
+def send_winter_storage_invoices(
+    request: HttpRequest, due_date: date = None
+) -> Dict[str, List]:
+    raise NotImplementedError
+
+
+@transaction.atomic
+def send_invoices(
+    request: HttpRequest,
+    lease_type: str,
+    season_start: date,
+    season_end: date,
+    due_date: date = None,
+) -> Dict[str, List]:  # noqa: C901
+    successful_orders: List[UUID] = []
+    failed_orders: List[Dict[UUID, str]] = []
+    failed_leases: List[Dict[UUID, str]] = []
+
+    # Default the due date to 14 days from the date when the task is executed
+    if not due_date:
+        due_date = today() + relativedelta(days=14)
+
+    leases = get_valid_leases(season_start, lease_type)
+
+    orders: List[Order] = []
+
+    for lease in leases:
+        lease = create_new_lease(lease, season_start, season_end)
+
+        # Get the associated product
+        product = get_product(lease)
+
+        # If no product is found, the billing can't be done
+        if not product:
+            fail_lease(lease, _("No suitable product found"), failed_leases)
```

Comment:
If no product is found, then I understand the lease is left in LeaseStatus.DRAFTED state. What happens to these leases afterwards? Is the draft lease in this case useful for admins, or alternatively should the draft be deleted? If leaving the draft lease in db is needed, maybe set also the comment field in lease to indicate the error?

Reply:
> then I understand the lease is left in LeaseStatus.DRAFTED state

The lease is set to `ERROR` status.

> maybe set also the comment field in lease to indicate the error

Yes! Great minds think alike, that's part of the improvements on the next PR